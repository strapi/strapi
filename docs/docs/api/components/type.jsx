import React from 'react';

export default function Type({ children }) {
  return (
    <span
      style={{
        color: '#017501',
      }}
    >
      &lt;{children}&gt;
    </span>
  );
}
