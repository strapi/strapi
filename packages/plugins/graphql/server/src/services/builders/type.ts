import { isArray, isString, isUndefined, constant } from 'lodash/fp';
import { nonNull, list, objectType } from 'nexus';
import { contentTypes } from '@strapi/utils';
import type { Struct } from '@strapi/types';

import type { Context } from '../types';

export type TypeBuildersOptions = {
  builder: any;
  attributeName: string;
  attribute: any;
  contentType: any;
  context: Context;
};

export default (context: Context) => {
  const { strapi } = context;
  const getGraphQLService = strapi.plugin('graphql').service;

  const extension = getGraphQLService('extension');

  /**
   * Add a scalar attribute to the type definition
   *
   * The attribute is added based on a simple association between a Strapi
   * type and a GraphQL type (the map is defined in `strapiTypeToGraphQLScalar`)
   */
  const addScalarAttribute = (options: TypeBuildersOptions) => {
    const { builder, attributeName, attribute } = options;

    const { mappers } = getGraphQLService('utils');

    const gqlType = mappers.strapiScalarToGraphQLScalar(attribute.type);

    builder.field(attributeName, { type: gqlType });
  };

  /**
   * Add a component attribute to the type definition
   *
   * The attribute is added by fetching the component's type
   * name and using it as the attribute's type
   */
  const addComponentAttribute = (options: TypeBuildersOptions) => {
    const { builder, attributeName, contentType, attribute } = options;

    let localBuilder = builder;

    const { naming } = getGraphQLService('utils');
    const { getContentTypeArgs } = getGraphQLService('builders').utils;
    const { buildComponentResolver } = getGraphQLService('builders').get('content-api');

    const type = naming.getComponentNameFromAttribute(attribute);

    if (attribute.repeatable) {
      localBuilder = localBuilder.list;
    }

    const targetComponent = strapi.getModel(attribute.component);

    const resolve = buildComponentResolver({
      contentTypeUID: contentType.uid,
      attributeName,
      strapi,
    });

    const args = getContentTypeArgs(targetComponent, {
      multiple: !!attribute.repeatable,
      isNested: true,
    });

    localBuilder.field(attributeName, { type, resolve, args });
  };

  /**
   * Add a dynamic zone attribute to the type definition
   *
   * The attribute is added by fetching the dynamic zone's
   * type name and using it as the attribute's type
   */
  const addDynamicZoneAttribute = (options: TypeBuildersOptions) => {
    const { builder, attributeName, contentType } = options;

    const { naming } = getGraphQLService('utils');
    const { ERROR_CODES } = getGraphQLService('constants');
    const { buildDynamicZoneResolver } = getGraphQLService('builders').get('content-api');

    const { components } = contentType.attributes[attributeName];

    const isEmpty = components.length === 0;
    const type = naming.getDynamicZoneName(contentType, attributeName);

    const resolve = isEmpty
      ? // If the dynamic zone don't have any component, then return an error payload
        constant({
          code: ERROR_CODES.emptyDynamicZone,
          message: `This dynamic zone don't have any component attached to it`,
        })
      : //  Else, return a classic dynamic-zone resolver
        buildDynamicZoneResolver({
          contentTypeUID: contentType.uid,
          attributeName,
        });

    builder.list.field(attributeName, { type, resolve });
  };

  /**
   * Add an enum attribute to the type definition
   *
   * The attribute is added by fetching the enum's type
   * name and using it as the attribute's type
   */
  const addEnumAttribute = (options: TypeBuildersOptions) => {
    const { builder, attributeName, contentType } = options;

    const { naming } = getGraphQLService('utils');

    const type = naming.getEnumName(contentType, attributeName);

    builder.field(attributeName, { type });
  };

  /**
   * Add a media attribute to the type definition
   */
  const addMediaAttribute = (options: TypeBuildersOptions) => {
    const { naming } = getGraphQLService('utils');
    const { getContentTypeArgs } = getGraphQLService('builders').utils;
    const { buildAssociationResolver } = getGraphQLService('builders').get('content-api');
    const extension = getGraphQLService('extension');

    const { builder } = options;
    const { attributeName, attribute, contentType } = options;
    const fileUID = 'plugin::upload.file';

    if (extension.shadowCRUD(fileUID).isDisabled()) {
      return;
    }

    const fileContentType = strapi.contentTypes[fileUID];

    const resolve = buildAssociationResolver({
      contentTypeUID: contentType.uid,
      attributeName,
      strapi,
    });

    const args = attribute.multiple
      ? getContentTypeArgs(fileContentType, { isNested: true })
      : undefined;

    const typeName = naming.getTypeName(fileContentType);

    if (attribute.multiple) {
      builder.field(`${attributeName}_connection`, {
        type: naming.getRelationResponseCollectionName(fileContentType),
        resolve,
        args,
      });

      builder.field(attributeName, {
        type: nonNull(list(typeName)),
        async resolve(...args: unknown[]) {
          const res = await resolve(...args);
          return res.nodes ?? [];
        },
        args,
      });
    } else {
      builder.field(attributeName, {
        type: typeName,
        async resolve(...args: unknown[]) {
          const res = await resolve(...args);
          return res.value;
        },
        args,
      });
    }
  };

  /**
   * Add a polymorphic relational attribute to the type definition
   */
  const addPolymorphicRelationalAttribute = (options: TypeBuildersOptions) => {
    const { GENERIC_MORPH_TYPENAME } = getGraphQLService('constants');
    const { naming } = getGraphQLService('utils');
    const { buildAssociationResolver } = getGraphQLService('builders').get('content-api');

    let { builder } = options;
    const { attributeName, attribute, contentType } = options;

    const { target } = attribute;
    const isToManyRelation = attribute.relation.endsWith('Many');

    if (isToManyRelation) {
      builder = builder.list;
    }
    // todo[v4]: How to handle polymorphic relation w/ entity response collection types?
    //  -> Currently return raw polymorphic entities

    const resolve = buildAssociationResolver({
      contentTypeUID: contentType.uid,
      attributeName,
      strapi,
    });

    // If there is no specific target specified, then use the GenericMorph type
    if (isUndefined(target)) {
      builder.field(attributeName, {
        type: GENERIC_MORPH_TYPENAME,
        resolve,
      });
    }

    // If the target is an array of string, resolve the associated morph type and use it
    else if (isArray(target) && target.every(isString)) {
      const type = naming.getMorphRelationTypeName(contentType, attributeName);

      builder.field(attributeName, { type, resolve });
    }
  };

  /**
   * Add a regular relational attribute to the type definition
   */
  const addRegularRelationalAttribute = (options: TypeBuildersOptions) => {
    const { naming } = getGraphQLService('utils');
    const { getContentTypeArgs } = getGraphQLService('builders').utils;
    const { buildAssociationResolver } = getGraphQLService('builders').get('content-api');
    const extension = getGraphQLService('extension');

    const { builder } = options;
    const { attributeName, attribute, contentType } = options;

    if (extension.shadowCRUD(attribute.target).isDisabled()) {
      return;
    }

    const isToManyRelation = attribute.relation.endsWith('Many');

    const resolve = buildAssociationResolver({
      contentTypeUID: contentType.uid,
      attributeName,
      strapi,
    });

    const targetContentType = strapi.getModel(attribute.target);

    const typeName = naming.getTypeName(targetContentType);

    const args = isToManyRelation
      ? getContentTypeArgs(targetContentType, { isNested: true })
      : undefined;

    const resolverScope = `${targetContentType.uid}.find`;
    const resolverPath = `${naming.getTypeName(contentType)}.${attributeName}`;

    extension.use({ resolversConfig: { [resolverPath]: { auth: { scope: [resolverScope] } } } });

    if (isToManyRelation) {
      builder.field(`${attributeName}_connection`, {
        type: naming.getRelationResponseCollectionName(targetContentType),
        resolve,
        args,
      });

      extension.use({
        resolversConfig: { [`${resolverPath}_connection`]: { auth: { scope: [resolverScope] } } },
      });

      builder.field(attributeName, {
        type: nonNull(list(typeName)),
        async resolve(...args: unknown[]) {
          const res = await resolve(...args);
          return res.nodes ?? [];
        },
        args,
      });
    } else {
      builder.field(attributeName, {
        type: typeName,
        async resolve(...args: unknown[]) {
          const res = await resolve(...args);
          return res.value;
        },
        args,
      });
    }
  };

  const isNotPrivate = (contentType: Struct.Schema) => (attributeName: string) => {
    return !contentTypes.isPrivateAttribute(contentType, attributeName);
  };

  const isNotDisabled = (contentType: Struct.Schema) => (attributeName: string) => {
    return extension.shadowCRUD(contentType.uid).field(attributeName).hasOutputEnabled();
  };

  return {
    /**
     * Create a type definition for a given content type
     * @param contentType - The content type used to created the definition
     * @return {NexusObjectTypeDef}
     */
    buildTypeDefinition(contentType: Struct.Schema) {
      const utils = getGraphQLService('utils');

      const { getComponentName, getTypeName } = utils.naming;
      const {
        isStrapiScalar,
        isComponent,
        isDynamicZone,
        isEnumeration,
        isMedia,
        isMorphRelation,
        isRelation,
      } = utils.attributes;

      const { attributes, modelType } = contentType;

      const attributesKey = Object.keys(attributes);

      const name = (modelType === 'component' ? getComponentName : getTypeName).call(
        null,
        contentType
      );

      return objectType({
        name,
        definition(t) {
          // add back the old id attribute on contentType if v4 compat is enabled
          if (
            modelType !== 'component' &&
            isNotDisabled(contentType)('id') &&
            strapi.plugin('graphql').config('v4CompatibilityMode', false)
          ) {
            t.nonNull.id('id', {
              deprecation: 'Use `documentId` instead',
            });
          }

          if (modelType === 'component' && isNotDisabled(contentType)('id')) {
            t.nonNull.id('id');
          }

          if (modelType !== 'component' && isNotDisabled(contentType)('documentId')) {
            t.nonNull.id('documentId');
          }

          if (strapi.plugin('graphql').config('v4CompatibilityMode', false)) {
            t.nonNull.field('attributes', {
              deprecation: 'Use root level fields instead',
              type: name,
              resolve: (parent) => parent,
            });

            t.nonNull.field('data', {
              deprecation: 'Use root level fields instead',
              type: name,
              resolve: (parent) => parent,
            });
          }

          /** Attributes
           *
           * Attributes can be of 7 different kind:
           * - Scalar
           * - Component
           * - Dynamic Zone
           * - Enum
           * - Media
           * - Polymorphic Relations
           * - Regular Relations
           *
           * Here, we iterate over each non-private attribute
           * and add it to the type definition based on its type
           */
          attributesKey
            // Ignore private attributes
            .filter(isNotPrivate(contentType))
            // Ignore disabled fields (from extension service)
            .filter(isNotDisabled(contentType))
            // Add each attribute to the type definition
            .forEach((attributeName) => {
              const attribute = attributes[attributeName];

              // We create a copy of the builder (t) to apply custom
              // rules only on the current attribute (eg: nonNull, list, ...)
              let builder: any = t;

              if (attribute.required) {
                builder = builder.nonNull;
              }

              const options: TypeBuildersOptions = {
                builder,
                attributeName,
                attribute,
                contentType,
                context,
              };

              // Enums
              if (isEnumeration(attribute)) {
                addEnumAttribute(options);
              }

              // Scalars
              else if (isStrapiScalar(attribute)) {
                addScalarAttribute(options);
              }

              // Components
              else if (isComponent(attribute)) {
                addComponentAttribute(options);
              }

              // Dynamic Zones
              else if (isDynamicZone(attribute)) {
                addDynamicZoneAttribute(options);
              }

              // Media
              else if (isMedia(attribute)) {
                addMediaAttribute(options);
              }

              // Polymorphic Relations
              else if (isMorphRelation(attribute)) {
                addPolymorphicRelationalAttribute(options);
              }

              // Regular Relations
              else if (isRelation(attribute)) {
                addRegularRelationalAttribute(options);
              }
            });
        },
      });
    },
  };
};
