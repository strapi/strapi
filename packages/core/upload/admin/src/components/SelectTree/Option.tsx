import * as React from 'react';

import { Flex, Typography } from '@strapi/design-system';
import { ChevronDown, ChevronUp } from '@strapi/icons';
import { useIntl } from 'react-intl';
import { components, OptionProps as ReactSelectOptionProps } from 'react-select';
import { styled } from 'styled-components';

import type { Folder } from '../../../../shared/contracts/folders';

const ToggleButton = styled(Flex)`
  align-self: flex-end;
  height: 2.2rem;
  width: 2.8rem;

  &:hover,
  &:focus {
    background-color: ${({ theme }) => theme.colors.primary200};
  }
`;

interface SelectProps {
  maxDisplayDepth: number;
  openValues: string[];
  onOptionToggle: (value: string) => void;
}

interface FolderWithDepth extends Folder {
  depth: number;
  value: string;
}

interface OptionProps extends ReactSelectOptionProps<FolderWithDepth, false> {
  selectProps: SelectProps & ReactSelectOptionProps<FolderWithDepth, false>['selectProps'];
}

export const Option = ({ children, data, selectProps, ...props }: OptionProps) => {
  const { formatMessage } = useIntl();
  const { depth, value, children: options } = data;
  const { maxDisplayDepth, openValues, onOptionToggle } = selectProps;
  const isOpen = openValues.includes(value);

  const Icon = isOpen ? ChevronUp : ChevronDown;

  return (
    <components.Option data={data} selectProps={selectProps} {...props}>
      <Flex alignItems="start">
        <Typography textColor="neutral800" ellipsis>
          <span style={{ paddingLeft: `${Math.min(depth, maxDisplayDepth) * 14}px` }}>
            {children}
          </span>
        </Typography>

        {options && options?.length > 0 && (
          <ToggleButton
            aria-label={formatMessage({
              id: 'app.utils.toggle',
              defaultMessage: 'Toggle',
            })}
            tag="button"
            alignItems="center"
            hasRadius
            justifyContent="center"
            marginLeft="auto"
            onClick={(event: React.MouseEvent<HTMLButtonElement>) => {
              event.preventDefault();
              event.stopPropagation();

              onOptionToggle(value);
            }}
          >
            <Icon width="1.4rem" fill="neutral500" />
          </ToggleButton>
        )}
      </Flex>
    </components.Option>
  );
};
