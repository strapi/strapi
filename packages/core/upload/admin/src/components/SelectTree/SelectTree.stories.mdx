import { useState } from 'react';
import { Meta, ArgsTable, Canvas, Story } from '@storybook/addon-docs';
import { SelectTree } from './SelectTree.tsx';

<Meta title="components/SelectTree" />

# SelectTree

This component is used in order to render a tree-structure as a Select. All options which can be passed
into `react-select` are forwarded.

## Usage

<Canvas>
  <Story name="base">
    {() => {
      const [value, setValue] = useState({ value: 22, label: 'Folder 2.2' });
      const options = [
        {
          value: 1,
          label: 'Folder 1',
        },
        {
          value: 2,
          label: 'Folder 2',
          children: [
            {
              value: 21,
              label: 'Folder 2.1',
            },
            {
              value: 22,
              label: 'Folder 2.2',
              children: [
                {
                  value: 221,
                  label: 'Folder 2.2.1',
                },
                {
                  value: 222,
                  label: 'Folder 2.2.2',
                },
              ],
            },
          ],
        },
        {
          value: 3,
          label: 'Folder 3',
        },
      ];
      return (
        <>
          <SelectTree
            options={options}
            defaultValue={value}
            onChange={({ value, label }) => setValue({ value, label })}
          />
          <p>Selected Value: {value.value}</p>
        </>
      );
    }}
  </Story>
</Canvas>

<ArgsTable of={SelectTree} />
