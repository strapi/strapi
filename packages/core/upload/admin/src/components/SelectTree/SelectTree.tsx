import * as React from 'react';

import { Box } from '@strapi/design-system';
import { Cross, CaretDown } from '@strapi/icons';
import ReactSelect, {
  components,
  GroupBase,
  StylesConfig,
  ClearIndicatorProps,
} from 'react-select';
import { styled, useTheme, DefaultTheme } from 'styled-components';

import { Option } from './Option';
import { flattenTree, FlattenedNode } from './utils/flattenTree';
import { getOpenValues } from './utils/getOpenValues';
import { getValuesToClose } from './utils/getValuesToClose';

const hasParent = (option: FlattenedNode<string | number | null>) => !option.parent;

export type OptionSelectTree = {
  value: string | number | null;
  label?: string;
  children?: OptionSelectTree[];
};

export interface SelectTreeProps<
  Option = unknown,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  IsMulti extends boolean = false,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  Group extends GroupBase<Option> = GroupBase<Option>,
> {
  maxDisplayDepth?: number;
  defaultValue?: {
    value?: string | number | null;
  };
  options: OptionSelectTree[];
  onChange?: (value: Record<string, string | number>) => void;
  name?: string;
  menuPortalTarget?: HTMLElement | null;
  inputId?: string;
  error?: string;
  ariaErrorMessage?: string;
  isDisabled?: boolean;
  disabled?: boolean;
}

export const SelectTree = ({
  options: defaultOptions,
  maxDisplayDepth = 5,
  defaultValue,
  ...props
}: SelectTreeProps) => {
  const flatDefaultOptions = React.useMemo(() => flattenTree(defaultOptions), [defaultOptions]);
  const optionsFiltered = React.useMemo(
    () => flatDefaultOptions.filter(hasParent),
    [flatDefaultOptions]
  );
  const [options, setOptions] = React.useState(optionsFiltered);
  const [openValues, setOpenValues] = React.useState(
    getOpenValues(flatDefaultOptions, defaultValue)
  );

  React.useEffect(() => {
    if (openValues.length === 0) {
      setOptions(flatDefaultOptions.filter((option) => option.parent === undefined));
    } else {
      const allOpenValues = openValues.reduce<(string | number | null)[]>((acc, value) => {
        const options = flatDefaultOptions.filter(
          (option) => option.value === value || option.parent === value
        );

        options.forEach((option) => {
          const values = getOpenValues(flatDefaultOptions, option);
          acc = [...acc, ...values];
        });

        return acc;
      }, []);

      const nextOptions = flatDefaultOptions.filter((option) =>
        allOpenValues.includes(option.value)
      );

      setOptions(nextOptions);
    }
  }, [openValues, flatDefaultOptions, optionsFiltered]);

  const handleToggle = (value: string | number | null) => {
    if (openValues.includes(value)) {
      const valuesToClose = getValuesToClose(flatDefaultOptions, value);
      setOpenValues((prev) => prev.filter((prevData) => !valuesToClose.includes(prevData)));
    } else {
      setOpenValues((prev) => [...prev, value]);
    }
  };

  return (
    <Select
      components={{ Option }}
      options={options}
      defaultValue={defaultValue}
      isSearchable={false}
      /* -- custom props, used by the Option component */
      maxDisplayDepth={maxDisplayDepth}
      openValues={openValues}
      onOptionToggle={handleToggle}
      /* -- / custom props */
      {...props}
    />
  );
};

type SelectOption = { value?: string | number | null; label?: string };

interface SelectProps<
  Option = SelectOption,
  IsMulti extends boolean = false,
  Group extends GroupBase<Option> = GroupBase<Option>,
> {
  components?: object;
  styles?: StylesConfig<Option, IsMulti, Group>;
  error?: string;
  ariaErrorMessage?: string;
  options: OptionSelectTree[];
  defaultValue?: {
    value?: string | number | null;
  };
  isSearchable?: boolean;
  maxDisplayDepth?: number;
  openValues?: (string | number | null)[];
  onOptionToggle?: (value: string | number | null) => void;
}

const Select = ({
  components = {},
  styles = {},
  error,
  ariaErrorMessage,
  ...props
}: SelectProps) => {
  const theme = useTheme();
  const customStyles = getSelectStyles(theme, error);

  return (
    <ReactSelect
      menuPosition="fixed"
      components={{
        ...components,
        ClearIndicator,
        DropdownIndicator,
        IndicatorSeparator: () => null,
        LoadingIndicator: () => null,
      }}
      aria-errormessage={error && ariaErrorMessage}
      aria-invalid={!!error}
      styles={
        { ...customStyles, ...styles } as StylesConfig<SelectOption, false, GroupBase<SelectOption>>
      }
      {...props}
    />
  );
};

const IconBox = styled(Box)`
  background: transparent;
  border: none;
  position: relative;
  z-index: 1;

  svg {
    height: 1.1rem;
    width: 1.1rem;
  }

  svg path {
    fill: ${({ theme }) => theme.colors.neutral600};
  }
`;

const ClearIndicator = (
  props: ClearIndicatorProps<SelectOption, false, GroupBase<SelectOption>>
) => {
  const Component = components.ClearIndicator;

  return (
    <Component {...props}>
      <IconBox tag="button" type="button">
        <Cross />
      </IconBox>
    </Component>
  );
};

const CarretBox = styled(IconBox)`
  display: flex;
  background: none;
  border: none;

  svg {
    width: 0.9rem;
  }
`;

const DropdownIndicator = ({ innerProps }: { innerProps: object }) => {
  return (
    <CarretBox paddingRight={3} {...innerProps}>
      <CaretDown />
    </CarretBox>
  );
};

const getSelectStyles = (
  theme: DefaultTheme,
  error?: string
): StylesConfig<SelectOption, false, GroupBase<SelectOption>> => {
  return {
    clearIndicator: (base: object) => ({ ...base, padding: 0, paddingRight: theme.spaces[3] }),
    container: (base: object) => ({
      ...base,
      background: theme.colors.neutral0,
      lineHeight: 'normal',
    }),
    control(base: object, state: { isFocused: boolean; isDisabled: boolean }) {
      let borderColor = theme.colors.neutral200;
      let boxShadowColor: string | undefined = undefined;
      let backgroundColor: string | undefined = undefined;

      if (state.isFocused) {
        borderColor = theme.colors.primary600;
        boxShadowColor = theme.colors.primary600;
      } else if (error) {
        borderColor = theme.colors.danger600;
      }

      if (state.isDisabled) {
        backgroundColor = `${theme.colors.neutral150} !important`;
      }

      return {
        ...base,
        fontSize: theme.fontSizes[2],
        height: 40,
        border: `1px solid ${borderColor} !important`,
        outline: 0,
        backgroundColor,
        borderRadius: theme.borderRadius,
        boxShadow: boxShadowColor ? `${boxShadowColor} 0px 0px 0px 2px` : '',
      };
    },
    indicatorsContainer: (base: object) => ({ ...base, padding: 0, paddingRight: theme.spaces[3] }),
    input: (base: object) => ({
      ...base,
      margin: 0,
      padding: 0,
      color: theme.colors.neutral800,
      gridTemplateColumns: '0 100%',
    }),
    menuPortal: (base: object) => ({
      ...base,
      zIndex: theme.zIndices.dialog,
      pointerEvents: 'auto',
    }),
    menu(base: object) {
      return {
        ...base,
        width: '100%',
        marginTop: theme.spaces[1],
        backgroundColor: theme.colors.neutral0,
        color: theme.colors.neutral800,
        borderRadius: theme.borderRadius,
        border: `1px solid ${theme.colors.neutral200}`,
        boxShadow: theme.shadows.tableShadow,
        fontSize: theme.fontSizes[2],
        zIndex: 2,
      };
    },
    menuList: (base: object) => ({
      ...base,
      paddingLeft: theme.spaces[1],
      paddingTop: theme.spaces[1],
      paddingRight: theme.spaces[1],
      paddingBottom: theme.spaces[1],
    }),
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    option(base: any, state: { isFocused: boolean; isSelected: boolean }) {
      let backgroundColor = base?.backgroundColor as string;

      if (state.isFocused || state.isSelected) {
        backgroundColor = theme.colors.primary100;
      }

      return {
        ...base,
        color: theme.colors.neutral800,
        lineHeight: theme.spaces[5],
        backgroundColor,
        borderRadius: theme.borderRadius,
        '&:active': {
          backgroundColor: theme.colors.primary100,
        },
      };
    },
    placeholder: (base: object) => ({
      ...base,
      color: theme.colors.neutral600,
      marginLeft: 0,
      overflow: 'hidden',
      textOverflow: 'ellipsis',
      whiteSpace: 'nowrap',
      maxWidth: '80%',
    }),
    singleValue(base: object, state: { isDisabled: boolean }) {
      let color = theme.colors.neutral800;

      if (state.isDisabled) {
        color = theme.colors.neutral600;
      }

      return { ...base, marginLeft: 0, color };
    },
    valueContainer: (base: object) => ({
      ...base,
      cursor: 'pointer',
      padding: 0,
      paddingLeft: theme.spaces[4],
      marginLeft: 0,
      marginRight: 0,
    }),
  };
};
