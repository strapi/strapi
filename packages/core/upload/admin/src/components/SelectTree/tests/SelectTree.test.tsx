import { DesignSystemProvider } from '@strapi/design-system';
import { act, render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';

import { SelectTree } from '../SelectTree';

import type { SelectTreeProps } from '../SelectTree';

const FIXTURE_OPTIONS = [
  {
    value: 'f-1',
    label: 'Folder 1',
  },

  {
    value: 'f-2',
    label: 'Folder 2',
    children: [
      {
        value: 'f-2-1',
        label: 'Folder 2-1',
      },

      {
        value: 'f-2-2',
        label: 'Folder 2-2',
        children: [
          {
            value: 'f-2-2-1',
            label: 'Folder 2-2-1',
          },
        ],
      },
    ],
  },
];

const ComponentFixture = (props: SelectTreeProps) => (
  <IntlProvider locale="en" messages={{}}>
    <DesignSystemProvider>
      <SelectTree defaultValue={{ value: 'f1' }} {...props} />
    </DesignSystemProvider>
  </IntlProvider>
);

const setup = (props?: SelectTreeProps) => {
  return new Promise((resolve) => {
    act(() => {
      resolve(render(<ComponentFixture {...props} options={FIXTURE_OPTIONS} />));
    });
  });
};

describe('SelectTree', () => {
  test('renders', async () => {
    expect(await setup()).toMatchSnapshot();
  });
});
