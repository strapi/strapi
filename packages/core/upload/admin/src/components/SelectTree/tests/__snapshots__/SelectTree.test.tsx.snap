// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectTree renders 1`] = `
{
  "asFragment": [Function],
  "baseElement": .c0 {
  padding-inline-end: 12px;
}

.c3 {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c1 {
  background: transparent;
  border: none;
  position: relative;
  z-index: 1;
}

.c1 svg {
  height: 1.1rem;
  width: 1.1rem;
}

.c1 svg path {
  fill: #666687;
}

.c2 {
  display: flex;
  background: none;
  border: none;
}

.c2 svg {
  width: 0.9rem;
}

<body>
    <div>
      <div
        class=" css-1iyvlxg-container"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-2-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
          role="log"
        />
        <div
          class=" css-oe0cnm-control"
        >
          <div
            class=" css-onz4xi-ValueContainer"
          >
            <div
              class=" css-1ryi7up-singleValue"
            />
            <input
              aria-activedescendant=""
              aria-autocomplete="list"
              aria-expanded="false"
              aria-haspopup="true"
              aria-invalid="false"
              aria-readonly="true"
              class="css-mohuvp-dummyInput-DummyInput"
              id="react-select-2-input"
              inputmode="none"
              role="combobox"
              tabindex="0"
              value=""
            />
          </div>
          <div
            class=" css-1hm8yhz-IndicatorsContainer"
          >
            <div
              aria-hidden="true"
              class="c0 c1 c2"
            >
              <svg
                fill="currentColor"
                height="16"
                viewBox="0 0 32 32"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m26.708 12.708-10 10a1 1 0 0 1-1.415 0l-10-10A1 1 0 0 1 6 11h20a1 1 0 0 1 .707 1.707"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <span
        class="c3"
      >
        <p
          aria-live="polite"
          aria-relevant="all"
          id="live-region-log"
          role="log"
        />
        <p
          aria-live="polite"
          aria-relevant="all"
          id="live-region-status"
          role="status"
        />
        <p
          aria-live="assertive"
          aria-relevant="all"
          id="live-region-alert"
          role="alert"
        />
      </span>
    </div>
  </body>,
  "container": .c0 {
  padding-inline-end: 12px;
}

.c3 {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c1 {
  background: transparent;
  border: none;
  position: relative;
  z-index: 1;
}

.c1 svg {
  height: 1.1rem;
  width: 1.1rem;
}

.c1 svg path {
  fill: #666687;
}

.c2 {
  display: flex;
  background: none;
  border: none;
}

.c2 svg {
  width: 0.9rem;
}

<div>
    <div
      class=" css-1iyvlxg-container"
    >
      <span
        class="css-1f43avz-a11yText-A11yText"
        id="react-select-2-live-region"
      />
      <span
        aria-atomic="false"
        aria-live="polite"
        aria-relevant="additions text"
        class="css-1f43avz-a11yText-A11yText"
        role="log"
      />
      <div
        class=" css-oe0cnm-control"
      >
        <div
          class=" css-onz4xi-ValueContainer"
        >
          <div
            class=" css-1ryi7up-singleValue"
          />
          <input
            aria-activedescendant=""
            aria-autocomplete="list"
            aria-expanded="false"
            aria-haspopup="true"
            aria-invalid="false"
            aria-readonly="true"
            class="css-mohuvp-dummyInput-DummyInput"
            id="react-select-2-input"
            inputmode="none"
            role="combobox"
            tabindex="0"
            value=""
          />
        </div>
        <div
          class=" css-1hm8yhz-IndicatorsContainer"
        >
          <div
            aria-hidden="true"
            class="c0 c1 c2"
          >
            <svg
              fill="currentColor"
              height="16"
              viewBox="0 0 32 32"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m26.708 12.708-10 10a1 1 0 0 1-1.415 0l-10-10A1 1 0 0 1 6 11h20a1 1 0 0 1 .707 1.707"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <span
      class="c3"
    >
      <p
        aria-live="polite"
        aria-relevant="all"
        id="live-region-log"
        role="log"
      />
      <p
        aria-live="polite"
        aria-relevant="all"
        id="live-region-status"
        role="status"
      />
      <p
        aria-live="assertive"
        aria-relevant="all"
        id="live-region-alert"
        role="alert"
      />
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
